src/puffin/src/file_format/reader/footer.rs

```diff
@@ -0,0 +1,261 @@
+            },
+            ParseStage::HeadMagic => BytesToRead {
+                offset: self.head_magic_offset(),
+                size: MAGIC.len() as _,
+            },
+            _ => unreachable!(),
+        };
+
+        Some(btr)
+    }
+
+    /// Processes the bytes that have been read according to the current parsing stage
+    /// and advances the parsing stage. It ensures the correct sequence of bytes is
+    /// encountered and stores the necessary information in the `StageParser`.
+    fn consume_bytes(&mut self, bytes: Vec<u8>) -> Result<()> {
+        match self.stage {
+            ParseStage::FootMagic => {
+                ensure!(bytes == MAGIC, MagicNotMatchedSnafu);
+                self.stage = ParseStage::Flags;
+            }
+            ParseStage::Flags => {
+                self.flags = Self::parse_flags(&bytes)?;
+                self.stage = ParseStage::PayloadSize;
+            }
+            ParseStage::PayloadSize => {
+                self.payload_size = Self::parse_payload_size(&bytes)?;
+                self.stage = ParseStage::Payload;
+            }
+            ParseStage::Payload => {
+                self.payload = self.parse_payload(bytes)?;
+                self.stage = ParseStage::HeadMagic;
+            }
+            ParseStage::HeadMagic => {
+                ensure!(bytes == MAGIC, MagicNotMatchedSnafu);
+                self.stage = ParseStage::Done;
+            }
+            _ => unreachable!(),
+        }
+
+        Ok(())
+    }
+
+    /// Finalizes the parsing process, ensuring all stages are complete, and returns
+    /// the parsed `FileMetadata`. It converts the raw footer payload into structured data.
+    fn finish(self) -> Result<FileMetadata> {
+        ensure!(
+            self.stage == ParseStage::Done,
+            ParseStageNotMatchSnafu {
+                expected: format!("{:?}", ParseStage::Done),
+                actual: format!("{:?}", self.stage),
+            }
+        );
+
+        // deserialize the footer payload
+        serde_json::from_slice(self.payload.as_slice()).context(DeserializeJsonSnafu)
+    }
+
+    fn parse_flags(bytes: &[u8]) -> Result<Flags> {
+        let n = u32::from_le_bytes(bytes.try_into().context(BytesToIntegerSnafu)?);
+        Ok(Flags::from_bits_truncate(n))
+    }
+
+    fn parse_payload_size(bytes: &[u8]) -> Result<u64> {
+        let n = i32::from_le_bytes(bytes.try_into().context(BytesToIntegerSnafu)?);
+        ensure!(n >= 0, UnexpectedFooterPayloadSizeSnafu { size: n });
+        Ok(n as u64)
+    }
+
+    fn parse_payload(&self, bytes: Vec<u8>) -> Result<Vec<u8>> {
+        // TODO(zhongzc): support lz4
+        ensure!(
+            !self.flags.contains(Flags::FOOTER_PAYLOAD_COMPRESSED_LZ4),
+            UnsupportedDecompressionSnafu {
+                decompression: "lz4"
+            }
+        );
+
+        Ok(bytes)
+    }
+
+    fn foot_magic_offset(&self) -> u64 {
+        self.file_size - 4
```

Comment:
Is it possible that the file size is less than 4, 8, 12, and so on?

Reply:
added some checked for it, PTAL